pil-analyzer/src/pil_analyzer.rs

```diff
@@ -174,6 +176,99 @@ impl PILAnalyzer {
         }
     }
 
+    /// Check that all match expressions are exhaustive.
+    pub fn match_exhaustiveness_check(&self) {
+        for (name, (_, value)) in &self.definitions {
+            let Some(value) = value else { continue };
+            for e in value.all_children() {
+                let patterns = if let Expression::MatchExpression(_, arms) = e {
+                    arms.iter()
+                        .map(|arm| PatternTuple {
+                            patterns: vec![arm.pattern.clone()],
+                        })
+                        .collect::<Vec<_>>()
+                } else {
+                    continue;
+                };
+                let witnesses = Self::usefulness(&patterns, Pattern::CatchAll.into());
+                if witnesses.is_empty() {
+                    panic!("Function {name} is not exhaustive")
+                }
+            }
+        }
+
+        for id in &self.identities {
+            let patterns = id
+                .all_children()
+                .filter_map(|e| {
+                    if let Expression::MatchExpression(_, arms) = e {
+                        Some(
+                            arms.iter()
+                                .map(|arm| PatternTuple {
+                                    patterns: vec![arm.pattern.clone()],
+                                })
+                                .collect::<Vec<_>>(),
+                        )
+                    } else {
+                        None
+                    }
+                })
+                .collect::<Vec<_>>();
+            for patterns in patterns {
+                let witnesses = Self::usefulness(&patterns, Pattern::CatchAll.into());
+                if witnesses.is_empty() {
+                    panic!("Identity is not exhaustive")
+                }
+            }
+        }
+    }
+
+    /// Check if a new pattern is useful in a set of patterns.
+    /// A pattern is useful if it is not covered by the other patterns.
+    /// If a pattern is useful, it is returned as a witness.
+    pub fn usefulness(patterns: &[PatternTuple], new_pattern: PatternTuple) -> Vec<PatternTuple> {
+        let mut witnesses = HashSet::new();
+
+        if new_pattern.patterns.is_empty()
+            || patterns.iter().any(|patterns| patterns.is_irrefutable())
+            || patterns.contains(&new_pattern)
+        {
+            return vec![];
+        }
+
+        let mut expanded_patterns = patterns.to_vec();
+        expanded_patterns.push(new_pattern.clone());
+        let constructors = expanded_patterns.clone();
+
+        for constructor in &constructors {
+            let specialized_new_pattern = if !new_pattern.is_empty() {
```

Comment:
Might be my personal preference but please avoid negation if possible (and swap the cases instead)